services/watson-extension/src/watson_extension/clients/insights/image_builder.py

```diff
@@ -0,0 +1,95 @@
+import abc
+from dataclasses import dataclass
+from typing import Optional
+
+import injector
+
+from watson_extension.clients import ContentSourcesURL
+from watson_extension.clients.identity import AbstractUserIdentityProvider
+from watson_extension.clients.platform_request import AbstractPlatformRequest
+
+
+@dataclass
+class EnableCustomRepositoriesResponse:
+    response: Optional[str] = None
+
+
+class ImageBuilderClient(abc.ABC):
+    @abc.abstractmethod
+    async def enable_custom_repositories(
+        self, version: str
+    ) -> EnableCustomRepositoriesResponse: ...
+
+
+class ImageBuilderClientHttp(ImageBuilderClient):
+    def __init__(
+        self,
+        content_sources_url: injector.Inject[ContentSourcesURL],
+        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
+        platform_request: injector.Inject[AbstractPlatformRequest],
+    ):
+        super().__init__()
+        self.content_sources_url = content_sources_url
+        self.user_identity_provider = user_identity_provider
+        self.platform_request = platform_request
+
+    async def enable_custom_repositories(
+        self,
+        version: str,
+    ) -> EnableCustomRepositoriesResponse:
+        request = "/api/content-sources/v1/popular_repositories/?offset=0&limit=20"
+        response = await self.platform_request.get(
+            self.content_sources_url,
+            request,
+            user_identity=await self.user_identity_provider.get_user_identity(),
+        )
+        response.raise_for_status()
+
+        content = await response.json()
+
+        repository = None
+        for repo in content["data"]:
+            if repo["suggested_name"] and repo["suggested_name"].startswith(version):
+                repository = repo
+                break
+
+        headers = {"Content-Type": "application/json"}
+        formatted = [
+            {
+                "name": repository["suggested_name"],
+                "distribution_arch": repository["distribution_arch"],
+                "distribution_versions": repository["distribution_versions"],
+                "gpg_key": repository["gpg_key"],
+                "metadata_verification": repository["metadata_verification"],
+                "snapshot": False,
+                "url": repository["url"],
+            }
+        ]
+
+        request = "/api/content-sources/v1.0/repositories/bulk_create/"
```

Comment:
Ideally, we should try to keep clients doing 1 request per function and returning the data as is whenever is possible (we don't need to parse it all, but parse what we really need).

The core should be using the client to fetch the data it needs, process it and preparing for the rendering. i.e. all the logic that VA requires should be there.

This will help us to test these pieces easily and prevent from large chunks of code that are harder to test.

Could you update the client to have these 2 methods and move the logic to the core?

Thanks!

Reply:
Just moved the logic to Core :+1: